chore: Add Windows helpers for non-ASCII texts

rolysr opened the pull request:
This PR introduces utility functions to handle non-ASCII text encoding on Windows platforms for printing and logging purposes. These utilities enable correct output of Unicode text when using `std::wcout` and `std::wcerr` streams, addressing limitations where standard streams could not properly display non-ASCII characters, even after setting the console code pages to UTF-8. The changes improve the reliability and correctness of console output for internationalized file paths, messages, and logs in SaunaFS tools and services. This PR is a helper step for supporting UTF-8 command line arguments, ensuring that command line inputs containing non-ASCII characters are correctly processed and displayed throughout the application.

rolysr committed:
chore: Add Windows helpers for non-ASCII texts

This commit introduces utility functions to handle non-ASCII text
encoding on Windows platforms for priting and logging purposes when
using the std::wcout and std::wcerr streams, which was not possible
before using the standard streams, even by setting console code pages
to UTF-8.

Signed-off-by: rolysr <rolysr@leil.io>

rolysr committed:
chore: Migrate to Windows console Unicode output

As new functions for printing Unicode messages to the Windows console
have been added, the intention is for now on to use these functions for
console output on Windows for ensuring support of non-ASCII characters.

Signed-off-by: rolysr <rolysr@leil.io>

rolysr committed:
fix(slogger): Fix slogger for non-ASCII on Windows

Previously if mountpoint identifier contained non-ASCII characters on
Windows client, slogger displayed garbled text in log messages and
its generated filename. This commit fixes the issue by ensuring proper
handling of non-ASCII characters in slogger generated file on Windows.

Signed-off-by: rolysr <rolysr@leil.io>

rolysr commented on src/mount/mount_info.cc:
```diff
@@ -135,41 +136,42 @@ std::string get_current_user_sid() {
 	if (!OpenProcessToken(GetCurrentProcess(), TOKEN_QUERY, &hToken)) {
-		std::wcerr << L"OpenProcessToken failed: " << GetLastError()
-		           << std::endl;
+		print_unicode_console_error(L"OpenProcessToken failed" + std::to_wstring(GetLastError()) +
```
Done.

rolysr commented on src/mount/mount_info.cc:
```diff
@@ -135,41 +136,42 @@ std::string get_current_user_sid() {
 	if (!pTokenUser) {
-		std::wcerr << L"Memory allocation failed" << std::endl;
+		print_unicode_console_error(L"Memory allocation failed: " +
+		                            std::to_wstring(GetLastError()) + L"\n");
```
Done.

rolysr commented on src/common/args_stat_encoding.h:
```diff
@@ -122,3 +147,25 @@ using saunafs_stat_t = struct stat;
+	char buf[4096];
+	size_t n;
+	while ((n = fread(buf, 1, sizeof(buf), fp)) > 0) { result.append(buf, n); }
+	fclose(fp);
```
Done.

rolysr commented on src/mount/mount_info.cc:
```diff
@@ -135,41 +136,42 @@ std::string get_current_user_sid() {
 
 	// Open the access token associated with the current process
 	if (!OpenProcessToken(GetCurrentProcess(), TOKEN_QUERY, &hToken)) {
-		std::wcerr << L"OpenProcessToken failed: " << GetLastError()
-		           << std::endl;
+		print_unicode_console_error(L"OpenProcessToken failed" + std::to_wstring(GetLastError()) +
+		                            L"\n");
 		return "";
```
Done.

rolysr commented on src/mount/mount_info.cc:
```diff
@@ -135,41 +136,42 @@ std::string get_current_user_sid() {
 
 	// Get the size of the user information in the token
 	if (!GetTokenInformation(hToken, TokenUser, nullptr, 0, &dwSize) &&
 	    GetLastError() != ERROR_INSUFFICIENT_BUFFER) {
-		std::wcerr << L"GetTokenInformation failed: " << GetLastError()
-		           << std::endl;
+		print_unicode_console_error(L"GetTokenInformation failed: " +
+		                            std::to_wstring(GetLastError()) + L"\n");
 		CloseHandle(hToken);
 		return "";
```
Done.

rolysr commented on src/mount/mount_info.cc:
```diff
@@ -135,41 +136,42 @@ std::string get_current_user_sid() {
 	if (!pTokenUser) {
-		std::wcerr << L"Memory allocation failed" << std::endl;
+		print_unicode_console_error(L"Memory allocation failed: " +
+		                            std::to_wstring(GetLastError()) + L"\n");
```
Done.

rolysr commented on src/mount/mount_info.cc:
```diff
@@ -135,41 +136,42 @@ std::string get_current_user_sid() {
 
 	// Retrieve the user information from the token
 	if (!GetTokenInformation(hToken, TokenUser, pTokenUser, dwSize, &dwSize)) {
-		std::wcerr << L"GetTokenInformation failed: " << GetLastError()
-		           << std::endl;
+		print_unicode_console_error(L"GetTokenInformation failed: " +
+		                            std::to_wstring(GetLastError()) + L"\n");
 		free(pTokenUser);
 		CloseHandle(hToken);
 		return "";
```
Done.

rolysr commented on src/mount/mount_info.cc:
```diff
@@ -135,41 +136,42 @@ std::string get_current_user_sid() {
 
 	// Convert the SID to a string
 	if (!ConvertSidToStringSid(pTokenUser->User.Sid, &pStringSid)) {
-		std::wcerr << L"ConvertSidToStringSid failed: " << GetLastError()
-		           << std::endl;
+		print_unicode_console_error(L"ConvertSidToStringSid failed: " +
+		                            std::to_wstring(GetLastError()) + L"\n");
 		free(pTokenUser);
 		CloseHandle(hToken);
 		return "";
```
Done.

rolysr commented on src/common/args_stat_encoding.h:
```diff
@@ -20,16 +20,41 @@
+	HANDLE hConsole = GetStdHandle(STD_ERROR_HANDLE);
+	DWORD written;
+	WriteConsoleW(hConsole, msg.c_str(), (DWORD)msg.size(), &written, nullptr);
+}
```
Considered.

rolysr commented on src/common/args_stat_encoding.h:
```diff
@@ -122,3 +147,25 @@ using saunafs_stat_t = struct stat;
+	std::string result;
+	char buf[4096];
+	size_t n;
+	while ((n = fread(buf, 1, sizeof(buf), fp)) > 0) { result.append(buf, n); }
```
Changed by another sager implementation

rolysr commented on src/common/args_stat_encoding.h:
```diff
@@ -122,3 +147,25 @@ using saunafs_stat_t = struct stat;
+inline bool read_file_to_string(const std::string &path, std::string &out) {
+#ifdef _WIN32
+	std::wstring wpath = utf8_to_wstring(path);
+	FILE *fp = _wfopen(wpath.c_str(), L"rb, ccs=UTF-8");
```
Changed implementation.

rolysr commented on src/common/args_stat_encoding.h:
```diff
@@ -20,16 +20,41 @@
+inline void print_unicode_console_error(const std::wstring &msg) {
+	HANDLE hConsole = GetStdHandle(STD_ERROR_HANDLE);
+	DWORD written;
+	WriteConsoleW(hConsole, msg.c_str(), (DWORD)msg.size(), &written, nullptr);
```
Considered, but for simplicity is better to do no prints on fallback.

rolysr commented on src/common/args_stat_encoding.h:
```diff
@@ -20,16 +20,41 @@
+inline void print_unicode_console_error(const std::wstring &msg) {
+	HANDLE hConsole = GetStdHandle(STD_ERROR_HANDLE);
+	DWORD written;
+	WriteConsoleW(hConsole, msg.c_str(), (DWORD)msg.size(), &written, nullptr);
```
Considered but for simplicity is better no prints on fallback.

uristdwarf commented on src/common/args_stat_encoding.h:
```diff
@@ -20,16 +20,46 @@
 #include <sys/stat.h>
+#include <utility>
 
 #ifdef _WIN32
-#include <shellapi.h>
 #include <windows.h>
-#include <string>
+#include <cstdint>
+#include <shellapi.h>
 #include <vector>
+#include <memory>
 
 using saunafs_stat_t = struct _stat64;
 
+// Converts a UTF-8 string to a wide string (UTF-16) on Windows.
+inline std::wstring utf8_to_wstring(const std::string &str) {
+	int wlen = MultiByteToWideChar(CP_UTF8, 0, str.c_str(), -1, nullptr, 0);
+	if (wlen <= 0) return L"";
+	std::wstring wstr(wlen, L'\0');
+	MultiByteToWideChar(CP_UTF8, 0, str.c_str(), -1, &wstr[0], wlen);
+	if (!wstr.empty() && wstr.back() == L'\0') wstr.pop_back();
+	return wstr;
+}
+
+inline void print_unicode_console(const std::wstring &msg) {
+	HANDLE hConsole = GetStdHandle(STD_OUTPUT_HANDLE);
+	DWORD written;
+	if (hConsole && hConsole != INVALID_HANDLE_VALUE) {
+		WriteConsoleW(hConsole, msg.c_str(), (DWORD)msg.size(), &written, nullptr);
+	}
+}
+
+inline void print_unicode_console_error(const std::wstring &msg) {
+	HANDLE hConsole = GetStdHandle(STD_ERROR_HANDLE);
+	DWORD written;
+	if (hConsole && hConsole != INVALID_HANDLE_VALUE) {
+		WriteConsoleW(hConsole, msg.c_str(), (DWORD)msg.size(), &written, nullptr);
+	}
+}
+
```
Wasn't there something similar for the saunafs command you implemented? Check for duplicate logic

rolysr replied:
In this case, for the `saunafs` command it was not needed to use this specific printing functions as it is handled by `fprintf` and only code page and argument conversion to UTF-8 was enough. In the case of the Windows client, most of prints are performed with the `std::wcout` and `std::wcerr` functions, which allow to print string wide text on Windows, making this visible for other applications like the SaunaFS Connect GUI. Using previous mentioned prints did not correctly keep consistent output with non-ASCII characters, so it was needed to created new utility functions for printing wide string text on Windows and supporting non-ASCII chars.

uristdwarf commented on src/common/args_stat_encoding.h:
```diff
@@ -122,3 +152,28 @@ using saunafs_stat_t = struct stat;
+
+// Reads the entire file into a std::string (UTF-8)
+inline bool read_file_to_string(const std::string &path, std::string &out) {
+#ifdef _WIN32
```
Couple of issues:

1. Reads can fail midway, even if the file can be opened (e.g EINTR, EIO etc.). They might be silently ignored or a exception throwed (not sure).
2. istreambuf_iterator is not really great for big files. Better for you to get the size of the file and then read everything at once.

Something like this:
```cpp

ifs.seekg(0, std::ios::end);
std::streampos size = ifs.tellg();
if (size < 0) { return false; }
ifs.seekg(0, std::ios::beg);

std::string result;
result.resize(static_cast<std::size_t>(size));
if (!size) {
    out.clear();
    return true;
}

ifs.read(result.data(), size);
if (ifs.bad() || (ifs.fail() && !ifs.eof())) {
    return false;
}
```

Also maybe the comment is a bit misleading, it doesn't actually do any checking/conversion on whether the file is UTF-8.

rolysr replied:
Done.

uristdwarf commented on src/common/args_stat_encoding.h:
```diff
@@ -122,3 +152,28 @@ using saunafs_stat_t = struct stat;
+	out = std::move(result);
+	return true;
+#else
+	std::ifstream ifs(path, std::ios::in | std::ios::binary);
```
Why not use ifstream in the WIN32? I don't care if you use C functions (as long as you check for errors, which you are currently not) if you think it's more performant/better, but it would better to use on or the other for both ifdefs if possible.

rolysr replied:
Done.

uristdwarf commented on src/metrics/CMakeLists.txt:
```diff
@@ -23,3 +23,8 @@ add_library(metrics ${METRICS_SOURCES})
+if(MINGW)
+  # Ignore sign-compare warnings on MinGW due to spdlog assertions on Windows
+  target_compile_options(metrics PRIVATE -Wno-error=sign-compare)
+endif()
```
Are you sure about ignoring this? Maybe try updating the version of spdlog on Windows, or looking for a issue/bug on this?

uristdwarf replied:
Also, metrics should not be required to be compiled on Windows.

rolysr replied:
In this case in tried latest spdlog available on vcpkg ports, also checking latest one available did not mention to solve this related check. On the other hand, I slightly changed the approach by adding corresponding flags on slogger.h in order to ignore sign comparison errors coming from the `spdlog` library only, not from any module from our code.

rolysr committed:
fix: Handle non-ASCII paths for I/O limits file

Due to recent changes on the Windows client handling of options to
support non-ASCII characters, the I/O limits configuration file path
needs to be opened using wide-character functions in order to avoid
potential issues with file access. This commit fixes that issue.

Signed-off-by: rolysr <rolysr@leil.io>

lgsilva3087 commented on src/common/args_stat_encoding.h:
```diff
@@ -20,16 +20,47 @@
+#include <fstream>
+#include <string>
 #include <sys/stat.h>
+#include <utility>
```
Looks unused, consider to remove it.

rolysr replied:
Done.

lgsilva3087 commented on src/common/args_stat_encoding.h:
```diff
@@ -20,16 +20,47 @@
+// Converts a UTF-8 string to a wide string (UTF-16) on Windows.
+inline std::wstring utf8_to_wstring(const std::string &str) {
+	int wlen = MultiByteToWideChar(CP_UTF8, 0, str.c_str(), -1, nullptr, 0);
+	if (wlen <= 0) return L"";
```
Use braces please.

rolysr replied:
Done.

lgsilva3087 commented on src/common/args_stat_encoding.h:
```diff
@@ -20,16 +20,47 @@
+	if (wlen <= 0) return L"";
+	std::wstring wstr(wlen, L'\0');
+	MultiByteToWideChar(CP_UTF8, 0, str.c_str(), -1, &wstr[0], wlen);
+	if (!wstr.empty() && wstr.back() == L'\0') wstr.pop_back();
```
Braces please.

rolysr replied:
Done.

lgsilva3087 commented on src/common/args_stat_encoding.h:
```diff
@@ -20,16 +20,47 @@
 using saunafs_stat_t = struct _stat64;
 
+// Converts a UTF-8 string to a wide string (UTF-16) on Windows.
+inline std::wstring utf8_to_wstring(const std::string &str) {
```
Consider adding air to the function to breathe, by adding empty line to separate logical blocks.

rolysr replied:
Done.

lgsilva3087 commented on src/mount/mount_info.cc:
```diff
@@ -20,6 +20,7 @@
 
 #include "common/platform.h"
 
+#include "common/args_stat_encoding.h"
```
Looks unused. If it is only used in WIND32, pleae wrap it.

rolysr replied:
Done.

lgsilva3087 commented on src/slogger/slogger.h:
```diff
@@ -27,17 +27,34 @@
 #include "common/syslog_defs.h"
```
Looks not used in Linux.

rolysr replied:
Done.

lgsilva3087 reviewed (commented):
Please see my suggestions.

lgsilva3087 reviewed (approved):
LGTM

rolysr committed:
feat: Add support for unicode when no console

Previosusly implemented versions for print_unicode_console and
print_unicode_console_error functions did not handle the case when
there was no console attached on Windows (for example, when output is
redirected to a file or when running as a background process). This
commit introduces new Windows API based implementation for managing
those scenarios.

Signed-off-by: rolysr <rolysr@leil.io>

rolysr commented on src/common/args_stat_encoding.h:
```diff
@@ -20,16 +20,66 @@
 using saunafs_stat_t = struct _stat64;
 
+// Converts a UTF-8 string to a wide string (UTF-16) on Windows.
+inline std::wstring utf8_to_wstring(const std::string &str) {
+	int wlen = MultiByteToWideChar(CP_UTF8, 0, str.c_str(), -1, nullptr, 0);
+
+	if (wlen <= 0) { return L""; }
+
+	std::wstring wstr(wlen, L'\0');
+	MultiByteToWideChar(CP_UTF8, 0, str.c_str(), -1, &wstr[0], wlen);
+
+	if (!wstr.empty() && wstr.back() == L'\0') { wstr.pop_back(); }
+
+	return wstr;
+}
```
Considered.

rolysr commented on src/common/args_stat_encoding.h:
```diff
@@ -122,3 +172,29 @@ using saunafs_stat_t = struct stat;
 #endif
+
+// Reads the entire file into a std::string (UTF-8)
+inline bool read_file_to_string(const std::string &path, std::string &out) {
+#ifdef _WIN32
+	std::filesystem::path fsPath{utf8_to_wstring(path)};
+#else
+	std::filesystem::path fsPath{path};
+#endif
+	std::ifstream ifs(fsPath, std::ios::binary);
+	if (!ifs) { return false; }
+
+	ifs.seekg(0, std::ios::end);
+	std::streampos size = ifs.tellg();
+	if (size < 0) { return false; }
+	ifs.seekg(0, std::ios::beg);
+
+	out.clear();
+	if (size == 0) { return true; }
+
+	out.resize(static_cast<std::size_t>(size));
+	if (!ifs.read(out.data(), size)) {
+		out.clear();
+		return false;
+	}
+	return true;
+}
```
is it not fully working on Windows side where is also needed.

rolysr commented on src/common/args_stat_encoding.h:
```diff
@@ -20,16 +20,66 @@
+	return wstr;
+}
+
+inline void winapi_print(const std::wstring &msg, bool is_error = false) {
+	HANDLE h = GetStdHandle(is_error ? STD_ERROR_HANDLE : STD_OUTPUT_HANDLE);
+	if (h == nullptr || h == INVALID_HANDLE_VALUE) return;
+
+	DWORD mode = 0;
+	BOOL isConsole = GetConsoleMode(h, &mode);
+
+	if (isConsole) {
+		// Console → Write UTF-16 directly
+		DWORD written;
+		WriteConsoleW(h, msg.c_str(), (DWORD)msg.size(), &written, nullptr);
+	} else {
+		// Pipe / file / redirection → must convert to UTF-8
+		int utf8len = WideCharToMultiByte(CP_UTF8, 0, msg.c_str(), (int)msg.size(), nullptr, 0,
+		                                  nullptr, nullptr);
+
+		if (utf8len <= 0) { return; }
+
+		std::string utf8(utf8len, '\0');
+
+		WideCharToMultiByte(CP_UTF8, 0, msg.c_str(), (int)msg.size(), utf8.data(), utf8len, nullptr,
+		                    nullptr);
+
+		DWORD written;
+		WriteFile(h, utf8.data(), (DWORD)utf8.size(), &written, nullptr);
+	}
+}
```
Considered.

rolysr commented on src/common/args_stat_encoding.h:
```diff
@@ -20,16 +20,66 @@
+#include <cstdint>
+#include <shellapi.h>
 #include <vector>
+#include <memory>
```
Done.

rolysr commented on src/mount/sauna_client.cc:
```diff
@@ -3636,15 +3637,16 @@ void fs_init(FsInitParams &params) {
-			if (!ifs.is_open()) {
+			std::string fileContent;
+			if (!read_file_to_string(params.io_limits_config_file, fileContent)) {
 				const char *strError = std::strerror(errno);
```
Done.

rolysr commented on src/common/args_stat_encoding.h:
```diff
@@ -122,3 +172,29 @@ using saunafs_stat_t = struct stat;
 inline int stat_portable(const char *path, saunafs_stat_t *st) { return stat(path, st); }
 #endif
+
+// Reads the entire file into a std::string (UTF-8)
```
Considered.

rolysr commented on src/common/args_stat_encoding.h:
```diff
@@ -20,16 +20,66 @@
+
+	return wstr;
+}
+
```
Done.

lgsilva3087 reviewed (approved):
LGTM. Great change @rolysr
